### ISSUE TITLE
Combine on corners of fields

### ISSUE THREAD

**Issue by BROADACRES16xFS17MAP** [Courseplay/courseplay#2242]
Hi, designing a map but combines keep getting stuck on the edges of some of the fields. Can you tell me the best way to fix it? Is it the terrain? I dont want huge bare patches between the fences and the field terrain as it looks weird. Any ideas to fix or improve?

First pic is were it gets stuck:

![fsscreen_2017_08_18_14_34_51](https://user-images.githubusercontent.com/7673205/29445272-0b0a22ea-8425-11e7-9fa9-cd21b0e5754b.png)

**Comment by gec100** [comment]
I've also noticed this problem.
One thing that could possibly be done is for the course generator to take in the calculation max radius of the machine.
Other thing would be make one more turn maneuver that would be used for corners. 

I do not know how to code so it's easy for me to say :)

**Comment by BROADACRES16xFS17MAP** [comment]
Thanks only fix at the moment is too make the corners rounded. Hopefully someone can give us an sight for a fix.

**Comment by pvajko** [comment]
The last comment at https://github.com/pvajko/course-generator/issues/13 summarizes the issue and the options.
I'm currently working on and making good progress on adding a Y turn at sharp corners on the headland. This solution is not using the turn system, it generates the turn during course generation (works with the new course generation only).

**Comment by BROADACRES16xFS17MAP** [comment]
Thanks so nothing is implemented so far?

**Comment by pvajko** [comment]
The new course generator uses 10% overlap on headlands which reduces the problem a bit. But you'll have to wait a few days for an alpha version of the new turns.

**Comment by BROADACRES16xFS17MAP** [comment]
Thanks obviously the 10% isn't enough in this case. Would you posting here once the alpha version is available? Happy to test. It will save me curving all my fields.

**Comment by pvajko** [comment]
Thanks, I'll let you kh9hnow here. 

Testing is the biggest problem as it takes a lot of time and people use all kinds of configurations, equipment, it's just impossible to test everything.

**Comment by BROADACRES16xFS17MAP** [comment]
Happy to help if you like. WE are currently in final testing for a 16x map so came into this issue once we commenced testing fields. Would great to have it fixed before release.
https://www.facebook.com/groups/1201506563300768/

**Comment by BROADACRES16xFS17MAP** [comment]
Any update on this one pvajko? :-)

**Comment by pvajko** [comment]
![turn2](https://user-images.githubusercontent.com/2379521/29592638-a4fc3158-87a5-11e7-892d-7285fb95db6d.PNG)

I was on a long transatlantic flight and hoped I can finish it but the plane had no power outlets...
Anyway, I  ended up implementing a new turn type, I have the geometry dialed in, what's missing is an option in the HUD to generate courses which have turn starts/ends when the direction change is big enough. I guess I won't be able to wrap it up before the weekend.

The current status is on the headlandTurns branch.

**Comment by BROADACRES16xFS17MAP** [comment]
Looks good mate, that would work, however what about when it has fences? Watching post with anticipation of a fix. :-)

**Comment by gec100** [comment]
yep. it should be going back than right in this situation. no? :)

**Comment by pvajko** [comment]
Well, this is similar to what the AIVE reverse with curve does but does not drive that far off the field. Do you have examples with fences really close to the field?

Anyway, I can do a straight reverse and then turn too if you guys think it's better.

**Comment by gec100** [comment]
coldborough park farm has fences really close if i remember correctly.

**Comment by BROADACRES16xFS17MAP** [comment]
Picture that i posted above is probably a good example of the fences.

**Comment by pvajko** [comment]
Oops, I did not notice that... Anyway:
![turn3](https://user-images.githubusercontent.com/2379521/29646025-8c8edebc-8846-11e7-88fc-6adc17f2df90.PNG)

**Comment by BROADACRES16xFS17MAP** [comment]
That looks better. Will that work if it gets caught on a 45 degree angle into the turn like ive pictured?

**Comment by pvajko** [comment]
It shouldn't get caught like that in the first place because it will back up far enough to be able to make the turn.

**Comment by BROADACRES16xFS17MAP** [comment]
Excellent lets run with that and see how it goes I guess. Happy to try it out if needing someone.

**Comment by BROADACRES16xFS17MAP** [comment]
Any update mate? No pressure at all :-) Just about to release my map for beta testing so not sure if I should make my corners round or hold of a bit for you? Cheers.

**Comment by pvajko** [comment]
The status is this: I have it working for combines, have an option in the HUD to generate the course with turns in the sharper corners instead of smoothing/rounding the corner when using the new course generator (starting location based on vehicle's location). It drives forward to the field edge and then reverses, and finally makes the turn. This works also for concave fields with 270 degree corners like 22 on Goldcrest. It will keep the combine on during the entire turn at the moment and won't raise the header.

The next steps would be raising/lowering implements within the turn (Courseplay can't do that at the moment) and to make it work for fieldwork like sowing where you would most likely want to take the turn first and then reverse.

These will have to wait now as fixing the existing problems has higher priority. 

The current status (combines) is available on GitHub on the headlandTurns branch (https://github.com/Courseplay/courseplay/tree/headlandTurns). If you want, you can give it a try and decide if you can live with it as it is. If yes, I can merge this to the master as a beta feature.

**Comment by BROADACRES16xFS17MAP** [comment]
Thanks mate i actually found the branch after i posted. I downloaded it and got the reverse part coming up in the hud however the header still got stuck on the turn as the first pic i initially sent and It wouldnt reverse.

**Comment by pvajko** [comment]
When you turn on the waypoints, do you see the green and red markers at the corner?

**Comment by pvajko** [comment]
Well, obviously this is not ready for beta then :)
@BROADACRES16xFS17MAP , if it turns out that this is a CP problem, I'd really love to have the map you have the issue with.

**Comment by BROADACRES16xFS17MAP** [comment]
I assume i am doing it right. I select reverse in the hud setup then generate the course with 3 headlands? The waypoints looked no different.

**Comment by pvajko** [comment]
You should see the waypoint before the corner green, the one at the corner red just like on the up/down rows marking the start and end of turns.

These are generated only when there's a direction change of 60 degrees or more between two subsequent waypoints. So if your field polygon's corners are a bit rounded it may not trigger the turn generation in the headland because it is based on the field boundary.

**Comment by BROADACRES16xFS17MAP** [comment]
Let me try again tonight. Is there a debug channel which will help us?

**Comment by pvajko** [comment]
Channel 7 and a 'cpSaveAllFields' command from the console which will create a cpFields.xml file under CoursePlay_Courses/<whatever your map is called> (you may need to save a course from that field before so the folder is created), I would need that file and the number of field.

**Comment by BROADACRES16xFS17MAP** [comment]
Okay retried on a big field but square. I have attached a link to the zipped debug files. Field 4 was the field. Hopefully I have done it right. No reverse attempted or waypoints shown.

https://mega.nz/#!ACx1ED7Z!0nNCcq5ogd3VElQFJSs8Osz7iw8Fx5QKEPReMKAFqx0



![fsscreen_2017_08_29_16_36_19](https://user-images.githubusercontent.com/7673205/29808855-ee0d6442-8cd8-11e7-997a-4d9c8a0c1033.png)



![fsscreen_2017_08_29_16_36_55](https://user-images.githubusercontent.com/7673205/29808871-0052ee56-8cd9-11e7-939c-ef653c83f469.png)

**Comment by pvajko** [comment]
Thx. My explanation probably wasn't clear, do you know how to start the game console and issue commands? Did you type in cpSaveAllFields in the game console? That should create the cpFields.xml file which I need.
What I see in the zip is that you saved a course name cpSaveAllFields.

**Comment by BROADACRES16xFS17MAP** [comment]
Haha u did wonder about that part. No u dint but happy to learn and try again.

**Comment by BROADACRES16xFS17MAP** [comment]
I did not u did

**Comment by pvajko** [comment]
Also, looking at your first screenshot, are you sure you were using the new course generator? Where the Starting location is 'Current vehicle position'?

**Comment by BROADACRES16xFS17MAP** [comment]
I had that option aswell as the reverse option so its definitely  the correct one. I didnt select the current vehicle position though. That needs to be done?

**Comment by pvajko** [comment]
You have to generate the course with the 'Current vehicle position' or 'Last used vehicle position' and the Generate headland turns set to 'Reverse straight' for this to work.

**Comment by BROADACRES16xFS17MAP** [comment]
No worries will test this afternoon.

**Comment by BROADACRES16xFS17MAP** [comment]
Okay so tested and it got what we needed however the reverse is still not long enough and still gets stuck on corners.

![fsscreen_2017_08_30_14_15_09](https://user-images.githubusercontent.com/7673205/29855910-82b80bbe-8d8e-11e7-8e12-919317559ecf.png)
![fsscreen_2017_08_30_14_15_29](https://user-images.githubusercontent.com/7673205/29855911-82b87518-8d8e-11e7-9a69-7718379bda9f.png)
![fsscreen_2017_08_30_14_15_52](https://user-images.githubusercontent.com/7673205/29855909-82b7f11a-8d8e-11e7-8093-bc73cba7caa9.png)

**Comment by BROADACRES16xFS17MAP** [comment]
When commencing the generation of the course the waypoint seems to be offline of were the start point of the combine is. Not sure if this may be the small issue. It appears slightly to the left of the centre of the combine which means the cutter is off the field to the left a bit.


![fsscreen_2017_08_30_14_25_50](https://user-images.githubusercontent.com/7673205/29856244-b94ab210-8d90-11e7-9a9d-47731d92104e.png)
![fsscreen_2017_08_30_14_31_29](https://user-images.githubusercontent.com/7673205/29856264-d46e7ebe-8d90-11e7-9746-52ea7fce1c08.png)

**Comment by pvajko** [comment]
The header is off the field a bit because I use an overlap in the headlands for various reason and due to a bug there was on overlap in the first headland too. I pushed a fixed for that to the headlandTurn branch.

It may also help to remedy the other problem but that needs a bit of a tuning, I noticed that the combine switches to the next waypoint too early, before it finishes the last curve.

Also, I need to find a better way to figure out how far forward we need to drive so the header reaches the edge of the field.

**Comment by BROADACRES16xFS17MAP** [comment]
No worries mate, downloaded new one and nil improvement. I guess back to the drawing board. Let us know once a fix or improvement comes out and we will retest.

**Comment by pvajko** [comment]
You have to regenerate the course to see the improvement.

**Comment by BROADACRES16xFS17MAP** [comment]
I got the improvement of the waypoints but it didnt help the reverse and cornering.

**Comment by BROADACRES16xFS17MAP** [comment]
Downloaded new update you did 7 hours ago. It now collects all the cotton to the fence but wants to keep going and not reverse.

![fsscreen_2017_09_01_17_16_56](https://user-images.githubusercontent.com/7673205/29960145-9a247bfa-8f39-11e7-84ee-238194ca79fe.png)

**Comment by BROADACRES16xFS17MAP** [comment]
In saying that just tested the next corner and its good.


![fsscreen_2017_09_01_17_35_33](https://user-images.githubusercontent.com/7673205/29960790-3d5ef8ca-8f3c-11e7-9648-114e1f64889e.png)
![fsscreen_2017_09_01_17_35_44](https://user-images.githubusercontent.com/7673205/29960791-3d73a72a-8f3c-11e7-933e-0fb615b768ae.png)

**Comment by BROADACRES16xFS17MAP** [comment]
But not on the third lol

**Comment by BROADACRES16xFS17MAP** [comment]
Any update at all mate?

**Comment by Willykeen** [comment]
Ive managed to re create the second last picture.
the problem is the gates are open therefore the header will not read a collision and will try to continue to make the turn. it works fine when gates are closed.

**Comment by BROADACRES16xFS17MAP** [comment]
Ah interesting, this could be something to be looked at pvajko? WIllykeen does it still turn okay aswell and take all the fruit up to the gate?

**Comment by Willykeen** [comment]
Yep! went right up to the gate then backed up and continued to turn

**Comment by pvajko** [comment]
Strange, I thought I reacted on this a few days ago, looks like it got lost somehow.

Anyway, it is really hard to figure out how far to go with a big combine like this and don't have more time to spend on this at the moment, so the easiest is probably leaving a little more room between the field and the fence, I understand this is a map you are building anyway.

I'll merge this state to the master as an experimental feature this week.

**Comment by pvajko** [comment]
This is now merged to the master branch (b109e0449933bf5f5dbffc0c56b4dca29952f98c ), pre-alpha stage.
Select Headland corners: Turn (combines only) in the Course generation menu to test it.

**Comment by BROADACRES16xFS17MAP** [comment]
Thanks has there been changes since last test or is this going to be it?

**Comment by pvajko** [comment]
What you can do (if you don't want to move the fence) is to play with 0.5 constant at line 1544 in turn.lua to control how far the combine drives forward before starting to reverse. 

The problem with these is things that what works for your field and combine may not work so well for others, or other folks have other preferences. That's the reason it's pre-alpha and I wanted to be it on the master to hear back from more people. So far I only have feedback from you and @Willykeen, seemingly both using the same map and combine.

**Comment by pvajko** [comment]
Closed. Will get back to this and add more functionality after the release.
